src/algorithms/tracking/AmbiguitySolver.cc

```diff
@@ -0,0 +1,266 @@
+#include <ActsExamples/EventData/Trajectories.hpp>
+#include <boost/container/flat_set.hpp>
+#include <boost/container/vector.hpp>
+#include <edm4eic/Cov6f.h>
+#include <edm4eic/Measurement2DCollection.h>
+#include <edm4eic/TrackParametersCollection.h>
+#include <edm4eic/TrajectoryCollection.h>
+#include <edm4hep/Vector2f.h>
+#include <edm4hep/Vector3f.h>
+#include <fmt/core.h>
+#include <Eigen/Core>
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <cstddef>
+#include <optional>
+#include <utility>
+
+#include "Acts/Utilities/Logger.hpp"
+#include "extensions/spdlog/SpdlogFormatters.h" // IWYU pragma: keep
+#include "extensions/spdlog/SpdlogToActs.h"
+
+namespace eicrecon {
+using namespace Acts::UnitLiterals;
+
+// This array relates the Acts and EDM4eic covariance matrices, including
+// the unit conversion to get from Acts units into EDM4eic units.
+//
+// Note: std::map is not constexpr, so we use a constexpr std::array
+// std::array initialization need double braces since arrays are aggregates
+// ref: https://en.cppreference.com/w/cpp/language/aggregate_initialization
+static constexpr std::array<std::pair<Acts::BoundIndices, double>, 6> edm4eic_indexed_units{
+    {{Acts::eBoundLoc0, Acts::UnitConstants::mm},
+     {Acts::eBoundLoc1, Acts::UnitConstants::mm},
+     {Acts::eBoundPhi, 1.},
+     {Acts::eBoundTheta, 1.},
+     {Acts::eBoundQOverP, 1. / Acts::UnitConstants::GeV},
+     {Acts::eBoundTime, Acts::UnitConstants::ns}}};
+
+Acts::GreedyAmbiguityResolution::Config
+transformConfig(const eicrecon::AmbiguitySolverConfig& cfg) {
+  Acts::GreedyAmbiguityResolution::Config result;
+  result.maximumSharedHits = cfg.maximum_shared_hits;
+  result.maximumIterations = cfg.maximum_iterations;
+  result.nMeasurementsMin  = cfg.n_measurements_min;
+  return result;
+}
+
+std::size_t sourceLinkHash(const Acts::SourceLink& a) {
+  return static_cast<std::size_t>(a.get<ActsExamples::IndexSourceLink>().index());
+}
+
+bool sourceLinkEquality(const Acts::SourceLink& a, const Acts::SourceLink& b) {
+  return a.get<ActsExamples::IndexSourceLink>().index() ==
+         b.get<ActsExamples::IndexSourceLink>().index();
+}
+
+AmbiguitySolver::AmbiguitySolver() {}
+
+void AmbiguitySolver::init(std::shared_ptr<spdlog::logger> log) {
+
+  m_log         = log;
+  m_acts_logger = eicrecon::getSpdlogLogger("AmbiguitySolver", m_log);
+  m_acts_cfg    = transformConfig(m_cfg);
+  m_core        = std::make_unique<Acts::GreedyAmbiguityResolution>(m_acts_cfg, logger().clone());
+}
+
+std::tuple<
+    std::unique_ptr<edm4eic::TrajectoryCollection>,
+    std::unique_ptr<edm4eic::TrackParametersCollection>, std::unique_ptr<edm4eic::TrackCollection>,
+    std::vector<ActsExamples::ConstTrackContainer*>, std::vector<ActsExamples::Trajectories*>>
+AmbiguitySolver::process(std::vector<const ActsExamples::ConstTrackContainer*> input_container,
+                         std::vector<const ActsExamples::Trajectories*> input_traj,
```

Comment:
I believe, we are looking to make this able to consume objects from our EDM4eic. Not sure how far that goal is.

Reply:
@wdconinc What is the suggestion here? In the current scenario, where we don't yet have Acts PODIO support available to us, would we push to do the back-and-forth conversions? That gives us immutability and serialization to disk. Do we get any gain in terms of maintainability towards your future effort?

Reply:
I assume this will become part of future project on using Acts PODIO. Marking resolved here.